### ISSUE TITLE
Add way to self-document metrics

### ISSUE THREAD

**Issue by jrcamp** [open-telemetry/opentelemetry-collector#985]
As we are adding more metrics-focused receivers we are adding many more metrics. Currently we describe these in code in in protobuf/internal representation and for documentation purposes must transcribe to markdown. This duplication results in additional effort to write and maintain and creates a risk that the code doesn't match the docs.

## Option 1 (Go)
Describe metrics in Go and derive metadata from there.

**Pros**
* The protobuf descriptions of the metrics already exist for the relevant receivers.

**Cons**
* Harder for non-Go programmers to read and modify.

## Option 2 (YAML)
Describe metrics and labels in external user-readable/writable file (e.g. yaml) that is then used to generate Go structures. This is the path we chose for SignalFx agent. We hand-write [metadata.yaml](https://github.com/signalfx/signalfx-agent/blob/master/pkg/monitors/diskio/metadata.yaml) files which results in [Go constants and helpers](https://github.com/signalfx/signalfx-agent/blob/master/pkg/monitors/diskio/genmetadata.go) being generated at build time.

**Pros**
* Better for writing free-form text than in Go (e.g. for long user-facing docs).
* Easy to consume by non-Go tooling since it's regular yaml.

**Cons**
* Requires extra build step (automated with `go generate`).

**Comment by bogdandrutu** [comment]
+1 on using YAML. If you can start contributing the generator code will be good to start defining at least the host metrics in YAML.

**Comment by jrcamp** [comment]
# Metric Metadata

## Overview

To keep things lean I'm proposing implementing things in phases. The first phase is the minimal needed while subsequent phases are more nice to haves. Possible phases:

Phase 1: define metadata format and generate Go constants from it at build time.
Phase 2: add support for declaring platforms (windows, linux, etc.) that various metrics and labels apply to.
Phase 3: add support for a more "fluent" generated API that removes additional boilerplate of using pdata directly.

Maybe we never even get to phase 2 and 3 but we can discuss them after phase 1 is done. I put it out there just to show long term possibilities.

## Phase 1

Each metric receiver will have a `metrics.yaml` file. In it will describe the labels and metrics emitted by that receiver. For example:

```yaml
name: hostmetricsreceiver
labels:
  device:
    # This label value is arbitrary.
    description: Device name such as /dev/sda1.
  direction:
    description: Indicates what type of operation.
    # Many times labels are a fixed set of known values.
    enum: [read, write]

metrics:
  host/filesystem/used:
    description: Filesystem bytes used.
    unit: bytes
    type: gauge/int64
    labels: [device]
```

At build time it would generate a file `generated_metrics.go` in the same package as the receiver. Example for the above:

```go
const typeStr configmodels.Type = "hostmetricsreceiver"

var metrics = struct {
  hostFilesystemUsed pdata.MetricDescriptor
}{
  func() pdata.MetricDescriptor {
    descriptor := pdata.NewMetricDescriptor()
    descriptor.InitEmpty()
    descriptor.SetName("host/filesystem/used")
    descriptor.SetDescription("Filesystem bytes used.")
    descriptor.SetUnit("bytes")
    descriptor.SetType(pdata.MetricTypeGaugeInt64)
    return descriptor
  }(),
}

var labels = struct {
  Device    string
  Direction string
}{"device", "direction"}

const (
  labelDirectionRead  = "read"
  labelDirectionWrite = "write"
)
```

Using it would then look something like:

```go
  m := pdata.NewMetric()
  metrics.hostFilesystemUsed.CopyTo(m.MetricDescriptor())
  dp := m.Int64DataPoints().At(0)
  dp.InitEmpty()
  dp.LabelsMap().Insert(labels.Device, "/dev/sda1")
  dp.LabelsMap().Insert(labels.Direction, labelDirectionWrite)
  dp.SetValue(23)
```

**Comment by james-bebbington** [comment]
This looks good to me. Keen to see a PoC 😄 

One other consideration is that it's possible to have two metrics that have a label with the same name but different content (e.g. cpu/usage "state" label vs memory/usage "state" label), so having labels being keyed by name could be problematic in rare cases.

**Suggestion for Phase 2:**

```yaml
name: hostmetricsreceiver
labels:
  state:
    os: [darwin, linux, windows] # default would be all
    description: State of Memory.
    # i.e. similar to the way go uses naming conventions to determine which files to include when building
    description_linux: State of Memory (Linux).
    description_windows: State of Memory (Windows).
    enum: [used, free]
    enum_linux: [used, free, buffered, cached, slab_reclaimable, slab_unreclaimable]
``` 
And then just create separate go files for each referenced OS:
```
generated_metrics.go
generated_metrics_darwin.go
generated_metrics_linux.go
generated_metrics_windows.go
```
